muxers/mplex/src/io.rs

```diff
@@ -0,0 +1,733 @@
+                                let below_limit = self.open_substreams.len() == self.config.max_substreams - 1;
+                                if below_limit {
+                                    ArcWake::wake_by_ref(&self.notifier_open);
+                                }
+                            },
+                            SubstreamState::Open => {
+                                debug!("Substream half-closed by remote: {:?}", stream_id);
+                                e.insert(SubstreamState::RecvClosed);
+                            },
+                        }
+                    }
+                }
+                Frame::Reset { stream_id } => {
+                    if let Some(state) = self.open_substreams.remove(&stream_id.into_local()) {
+                        debug!("Substream {:?} in state {:?} reset by remote", stream_id, state);
+                        let below_limit = self.open_substreams.len() == self.config.max_substreams - 1;
+                        if below_limit {
+                            ArcWake::wake_by_ref(&self.notifier_open);
+                        }
+                    } else {
+                        trace!("Received reset for unknown stream: {:?}", stream_id);
+                    }
+                }
+                _ => ()
+            }
+
+            if let Some(frame) = select(&frame) {
+                // The caller is interested in this frame.
+                return Poll::Ready(Ok(Some(frame)));
+            } else {
+                // The caller is not interested in this frame, so we need to buffer
+                // it and wake those tasks who are interested.
+                let id = frame.local_id();
+                if self.can_read(&id) || frame.is_open() {
+                    trace!("Buffering frame: {:?} (total: {})", frame, self.buffer.len() + 1);
+                    self.buffer.push(frame);
+                    self.notifier_read.wake_by_id(id);
+                } else if frame.is_data() {
+                    debug!("Dropping {:?} for closed or unknown substream: {:?}", frame, id);
+                }
+            }
+        }
+    }
+
+    /// Generates the next outbound stream ID.
+    fn next_outbound_stream_id(&mut self) -> LocalStreamId {
+        let id = self.next_outbound_stream_id;
+        self.next_outbound_stream_id = self.next_outbound_stream_id.next();
+        id
+    }
+
+    /// Checks whether a substream is open for reading.
+    fn can_read(&self, id: &LocalStreamId) -> bool {
+        match self.open_substreams.get(id) {
+            Some(SubstreamState::Open) | Some(SubstreamState::SendClosed) => true,
+            _ => false,
+        }
+    }
+
+    /// Sends pending frames, without flushing.
+    fn send_pending_frames(&mut self, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        while let Some(frame) = self.pending_frames.pop() {
+            if self.poll_send_frame(cx, || frame.clone())?.is_pending() {
```

Comment:
Why not call `self.pending_frames.pop()` within the closure? As far as I understand, if the closure is called, the outcome cannot be `Pending`.

Reply:
I don't think that's permissible due to borrowing. Even if `poll_send_frame` were made a standalone function and thus not borrow all of `self`, it still calls `on_error` in case of an I/O error which also wants a mutable borrow for `pending_frames` to drop/clear it on error. I agree that the pop/push as well as the frame cloning is suboptimal, but I didn't consider it that important or inefficient. The frames sent here are only close/reset frames containing only a stream ID. We could of course make `poll_send_frame` return the frame in the `Pending` case with a custom return enum.